dracut/03flatcar-network/parse-ip-for-networkd.sh

```diff
@@ -1,17 +1,58 @@
-#!/bin/sh
+#!/bin/bash
 #
 # This script was mostly stolen from 40network/parse-ip-opts.sh.  Its
 # actions are adapted to write .network files to /etc/systemd/network
 # in the initramfs instead of using separate DHCP commands, etc.  Note
-# the bashisms.
+# the bashisms. It used to run as dracut hook, relying on afterburn to
+# process kargs for the dracut cmdline target but since that is too
+# early now that we start afterburn from /sysusr/ this script here
+# became an own systemd unit running shortly before systemd-networkd.
+# However, since the "netroot" variable is set up by another dracut
+# hook, this script here will still get executed once as dracut hook
+# to save the "netroot" variable to an env file and exit.
+# Could the script instead be a hook at a later stage? There doesn't
+# seem to be a suitable later stage and also then the env could be
+# different. As long as we want to rely on dracut-lib/net-lib for
+# parsing the env var topic is not easily avoided.
 #
 
+if [ "${APPLY-}" != "1" ]; then
+  # First run, called as dracut hook
+  {
+  # While the script only makes use of the env vars "NEEDBOOTDEV" and "netroot"
+  # we actually don't really know what env vars dracut-lib.sh and net-lib.sh
+  # will depend on in future versions and therefore, we try to preserve the
+  # environment as is.
+  for VARNAME in $(compgen -v); do
+    # Prevent leaking HEREEOF into VAL
+    [ "${VARNAME}" = "_" ] && continue
```

Comment:
Could you explain what's going on here, please? I don't get it. :)

Reply:
`$_` is the last argument of the previous command and it will result in `HEREEOF` being leaked into the HEREDOC.
E.g., if you run `echo "X=\$(cat <<'HEREEOF' ; Y=$_ ; HEREEOF ; )"` multiple times in a row the first run will insert what you ran before but if you run it again and again multiple HEREEOFs will be inserted recursively.

Reply:
Added it as explanation